state/optimistic_syncing/pipeline.go

```diff
@@ -0,0 +1,474 @@
+func (p *Pipeline) SetSealed() {
+	p.mu.Lock()
+	p.isSealed = true
+	p.mu.Unlock()
+
+	// Trigger state check
+	p.stateNotifier.Notify()
+}
+
+// updateState updates the internal state from a parent update and returns whether
+// the pipeline should be abandoned (no longer descends from latest sealed)
+func (p *Pipeline) updateState(update StateUpdate) bool {
+	p.mu.Lock()
+	previousDescendsFromLatest := p.descendsFromSealed
+	p.descendsFromSealed = update.DescendsFromLastPersistedSealed
+	p.parentState = update.ParentState
+	p.mu.Unlock()
+
+	return previousDescendsFromLatest && !update.DescendsFromLastPersistedSealed
+}
+
+// listenForStateUpdates listens for state updates from the parent pipeline
+// and updates internal state accordingly. If the pipeline no longer descends
+// from the latest persisted result, it transitions to the StateCanceled.
+func (p *Pipeline) listenForStateUpdates(ctx context.Context, cancelFunc context.CancelFunc) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case update := <-p.stateUpdateChan:
+			if shouldAbandon := p.updateState(update); shouldAbandon {
+				p.broadcastStateUpdate()
+				cancelFunc()
+			} else {
+				// Trigger state check
+				p.stateNotifier.Notify()
+			}
+		}
+	}
+}
+
+// broadcastStateUpdate sends a state update to all children pipelines.
+// The update includes the current pipeline state and whether it descends
+// from the latest persisted result.
+func (p *Pipeline) broadcastStateUpdate() {
+	if len(p.childrenUpdateChans) == 0 {
+		return
+	}
+
+	p.mu.RLock()
+	update := StateUpdate{
+		DescendsFromLastPersistedSealed: p.descendsFromSealed,
+		ParentState:                     p.state,
+	}
+	channels := p.childrenUpdateChans
+	p.mu.RUnlock()
+
+	for _, ch := range channels {
+		ch <- update
+	}
+}
+
+// processCurrentState handles the current state and transitions to the next state if possible.
+// It returns false when a terminal state is reached (StateComplete or StateCanceled), true otherwise.
+// Returns an error if any processing step fails.
+func (p *Pipeline) processCurrentState(ctx context.Context) (bool, error) {
+	currentState := p.GetState()
+
+	switch currentState {
+	case StateReady:
+		return p.processReady(), nil
+	case StateDownloading:
+		return p.processDownloading(ctx)
+	case StateIndexing:
+		return p.processIndexing(ctx)
+	case StateWaitingPersist:
+		return p.processWaitingPersist(), nil
+	case StatePersisting:
+		return p.processPersisting(ctx)
+	case StateComplete, StateCanceled:
+		// Terminal states
+		return false, nil
+	default:
+		return false, fmt.Errorf("%w: %s", ErrInvalidState, currentState.String())
+	}
+}
+
+// transitionTo transitions the pipeline to the given state and broadcasts
+// the state change to children pipelines.
+func (p *Pipeline) transitionTo(newState State) {
+	p.mu.Lock()
+	oldState := p.state
+	p.state = newState
+	p.mu.Unlock()
+
+	p.logger.Info().
+		Str("oldState", oldState.String()).
```

Comment:
same comment about snake case. here and all other labels
```suggestion
		Str("old_state", oldState.String()).
```